include/spl/sys/kmem_cache.h

```diff
@@ -43,9 +43,10 @@ enum {
 	KMC_BIT_QCACHE		= 4,	/* XXX: Unsupported */
 	KMC_BIT_KMEM		= 5,	/* Use kmem cache */
 	KMC_BIT_VMEM		= 6,	/* Use vmem cache */
-	KMC_BIT_SLAB		= 7,	/* Use Linux slab cache */
-	KMC_BIT_OFFSLAB		= 8,	/* Objects not on slab */
-	KMC_BIT_NOEMERGENCY	= 9,	/* Disable emergency objects */
+	KMC_BIT_KVMEM		= 7,	/* Use kvmalloc linux allocator  */
```

Comment:
What is the benefit of introducing this flag? Are there any consumers that would benefit from it? If not, we should just drop it and revisit the idea when there is an actual need.

Quite frankly, there really is not much difference between kmem and vmem from the perspective of a consumer. They are both mapped via page tables. The only difference is that the page tables mappings for kmem are static and mapped to the actual physical addresses while the page table mappings for vmem are dynamic. We are free to reinterpret KMC_BIT_VMEM to mean KMC_BIT_KVMEM. It makes little difference to the caller unless the caller is unmapping the pages being given to it. That would be a bug.

Reply:
The reason for this is explained here: https://github.com/zfsonlinux/zfs/pull/9034/files/ed7818dedd88e54c72600a3b5d4df08770a0918e#diff-57deeea9818bf198fc721d5486bb1982R209

TL;DR performance

Reply:
Why not change the semantics of KMC_BIT_VMEM to those of KMC_BIT_KVMEM?

Reply:
On a related topic, regardless of whether we introduce a third flag or keep only two, we really should add assertions to the code to verify that no more than 1 of these flags is ever set at a time.

Reply:
> Why not change the semantics of KMC_BIT_VMEM to those of KMC_BIT_KVMEM? I am not convinced that we need three different flags for controlling the allocation behavior. Two seems good enough unless there is a reason not to abandon the old semantics of KMC_BIT_VMEM.

We did not agree, yet, if kvmalloc can be used as a replacement for any vmalloc allocations if kvmalloc is available. If this is possible, we surely can drop KMC_BIT_KVMEM and translate/interpret KMC_BIT_VMEM.

Reply:
Let's leave this point open for discussion for now and first get this working - then we can decide what is better.

Reply:
I guess the question would be answered by determining whether kvmalloc() permits allocations greater than PAGE_SIZE << MAX_ORDER. The official kernel documentation suggests that it does:

https://www.kernel.org/doc/html/latest/core-api/memory-allocation.html

Reply:
I would rather tackle this as part of the patch to have a cleaner codebase when it is merged, rather than as a cleanup task for later. Simpler code also makes review easier.

Reply:
Ah, I just remembered what @BrainSlayer said in the discussion about that KMC bit above here https://github.com/zfsonlinux/zfs/pull/9034#issuecomment-513142474

This will be needed for zstd (#8044, #8941, #9024) since without forcing k(v)malloc, performance of zstd would decrease.

Reply:
If we use `kvalloc()` in zstd, we can end up with the system blocking on insufficient memory in the ZIO pipeline, deadlocking the system. I don't think this is a valid reason to avoid unifying the two.

Reply:
Also, `spl_kmem_alloc_max` wouldn't apply to `KM_VMEM` if it is being redefined to mean `kvmalloc()`.